lib/template-loader.php

```diff
@@ -230,110 +156,35 @@ function gutenberg_find_template_post_and_parts( $template_type, $template_hiera
 		$template_hierarchy
 	);
 
-	// Find most specific 'wp_template' post matching the hierarchy.
+	// Find all potential templates 'wp_template' post matching the hierarchy.
 	$template_query = new WP_Query(
 		array(
 			'post_type'      => 'wp_template',
-			'post_status'    => 'publish',
+			'post_status'    => array( 'publish', 'auto-draft' ),
 			'post_name__in'  => $slugs,
 			'orderby'        => 'post_name__in',
-			'posts_per_page' => 1,
+			'posts_per_page' => -1,
 			'no_found_rows'  => true,
+			'meta_key'       => 'theme',
+			'meta_value'     => wp_get_theme()->get( 'TextDomain' ),
```

Comment:
Also `wp_get_theme()->get_stylesheet()` 🙂

Reply:
Why should template parts have "theme" and not templates? There's no difference for me and it's the only way to know which template to export. 

What happens when you switch themes?  Do you want to keep your custom templates created based on the previous theme? That sounds confusing?

Reply:
Oh it is **very** confusing!

Is it more confusing to replace a custom template with a theme-provided one when activating a new theme?
Or is it more confusing to keep the custom template, which style might not match the new theme?

We've been doing lots of back and forth, and my understanding is that, at least for now, we should not have multiple templates for the same type.
(E.g. we should only have one `index`, rather than one from 2020, one from 2021, one custom, etc.)

This by itself would solve the export confusion: just export everything.

I'm personally a fan of having multiple templates, so that the user can easily swap them, instead of having to redesign them. But for that we would also need some UI to let the user select, for example, which `index` to use among those available.

I've opened an issue to discuss this, so it's not like any decisions is final yet. Feel free to chime in there!
https://github.com/WordPress/gutenberg/issues/26546

Reply:
Unrelated to the discussion, there's also another issue about using meta for this kind of data: https://github.com/WordPress/gutenberg/issues/24377

The tl;dr would be that we should use a taxonomy rather than a meta for performance reasons.
There is also a PR open about it that we should probably review ASAP: https://github.com/WordPress/gutenberg/pull/24720

Reply:
> Or is it more confusing to keep the custom template, which style might not match the new theme?

I believe this one is the most confusing since it's different from the current theme switch behavior.

I agree that this assumption can be challenged but I don't see why it would be different between templates and template parts. I think for the moment I'm keeping it (in this PR) but I'm open to explorations to remove these, but it should be removed for template parts too in that case.

Reply:
I also believe the templates that are not from the current theme should be entirely hidden in the site editor for now.

Reply:
> I agree that this assumption can be challenged but I don't see why it would be different between templates and template parts.

Im a bit torn between the argument of theme meta vs not for templates as well.  But I also think templates/template-parts should be consistent in this.

> The current approach we are evaluating is to wipe all auto-draft templates (which are those provided by the theme, that haven't been edited by the user) on theme change.

I am assuming if we wanted template parts to be consistent with not having a theme meta, we would need to do something like this as well.  However, in the current implementation you can add a template-part auto-draft wherever you want and it will stay an auto-draft as long as it is not edited.  Imagine a user likes a template part supplied by a theme so much that they add it to a custom template(s) and never edit the content of the template part, if auto-draft template parts were wiped on theme change then that custom template would lose part of its expected content. 🤔   That scenario right there makes it difficult for template parts (as they currently work) to follow that same sort of behavior as suggested for templates.

Reply:
It seems like the desire to be consistent for the sake of being consistent is getting in the way of pragmatism. I think there's a very clear case to use a theme meta for template parts but not for templates, and it's already just been explained by @Addison-Stavlo:

> Imagine a user likes a template part supplied by a theme so much that they add it to a custom template(s) and never edit the content of the template part, if auto-draft template parts were wiped on theme change then that custom template would lose part of its content.

This same scenario just doesn't exist for templates. We can safely say that an unedited auto-draft template will not be invoked after switching themes, because templates can't get embedded inside other content and the user hasn't made any edits that need to be stored. Therefore, there's no need to keep unedited auto-draft templates from inactive themes.

On the question of whether to keep edits to templates from the previous theme, I would look at it from this perspective:

- If the user's edits are kept and they don't want them, they can manually delete their own template and the new theme will automatically take over. The worst that happens is it causes the user a little bit of extra work to get their site looking the way they want.

- If the user's edits are not kept, how does the user reinstate them if they do want to keep them? Not only is it unclear how the UI would expose this, but it's also a betrayal of the user by applying changes they don't want. I think the other option is safer.

Furthermore, sites don't change their themes frequently and almost never reactivate old themes that were previously active. Keeping edited templates from previously active themes just seems to be going overboard. Maybe this could be an enhancement to a later version of FSE, but I strongly think the initial release of FSE should keep it simple and leave the theme meta to template parts only.

Reply:
Also, I would point out that no one asked for a theme meta on templates. What bug does it fix? It seems like a solution looking for a problem. Isn't it better to stick with the status quo until a good reason is found to change the database schema?

Reply:
> This same scenario just doesn't exist for templates

Why do you think so? I have no doubt that for instance a user might like the "contact page" from a theme and want to use it in another one?

I do think ultimately, "theme" doesn't matter and we can even go theme-less but I believe we want to go in steps. With the meta, we're matching the behavior of non-fse themes for theme switching which is the less confusing for existing users.

I agree thought that with block templates we have a lot more options to explore that were just not possible with existing themes. Basically we can mix templates and templates parts from any theme. This force us to think about this differently. Instead of assuming that folks would want to reuse template parts but not templates, I think there's a wide range of possibilities. What if the decision comes down to the user instead, what if there's a UI on theme switch to allow the user to pick what he wants, or what if we keep all templates and templates parts (by having the meta like now) and build UIs in templates and template parts loading and saving flows for the user to decide.

This lead me to think that until we have some design explorations on how we want this to work exactly, we'd want to match the existing behavior of themes while allowing all this range of options to be explored.

Reply:
> With the meta, we're matching the behavior of non-fse themes for theme switching which is the less confusing for existing users.

We might not be on the same page here. I think the theme meta does exactly the opposite, it departs from the behaviour of non-FSE themes. Leaving out the theme meta matches the behaviour of non-FSE themes, the way I see it.

If you want an example, I would point you to the one you just gave:

> I have no doubt that for instance a user might like the "contact page" from a theme and want to use it in another one?

I think what you're saying here is carry over the template without making any manual edits and inherit the theme updates, thereby mixing and matching unedited templates from two themes. This is a departure from non-FSE behaviour that the theme meta would enable.

As I said in my previous comment, this feature (and hence the theme meta that enables it) might be a candidate for a later version of FSE.

But, if you agree that the initial release should start by matching the behaviour of non-FSE themes (which it seems you do) then I don't see any reason the theme meta is needed at this time.

Reply:
> Instead of assuming that folks would want to reuse template parts but not templates

It's not about whether they want it, it's about whether it's necessary to prevent sites from breaking. The current implementation allows users to request a specific template part from a specific theme, but it doesn't allow them to do the same with templates. Whether that feature should be added is probably not a question for this pull request to be prematurely answering?

Reply:
If I'm reading this right, I think we agree that we should match existing themes switch but we defer on implementation (which is fine, it's just implementation detail anyway).

1- **master** theme switch work like non-fse themes from templates but doesn't for template parts. (you could end up with a header template part from a theme used on another theme)

2- **PR** theme switch work like non-fse themes from both templates and template parts, the difference is that edited templates and template parts are kept on the database but not used in the rendering (thanks to the meta flag). Switching back to the old theme will enable them again.

3- **alternative proposal 1** Ditch the theme meta entirely, this means mixing templates and templates from different themes on switch. (doesn't match non-FSE themes)

4- **alternative proposal 2** Ditch the theme meta entirely but also remove edited templates and template parts on theme switch (match non-FSE themes)

---
Right now, I can be on board with 2 and 4 (to match non-FSE themes). (this is what is acceptable for me to land this PR) 

Ultimately, I think we should do **2** but enable flows in the UI to mix template and template parts with conscious decisions from the user instead of making arbitrary decisions. I think this should be explored in terms of design first.

Reply:
It suddenly hit me what I was missing. Referring again to this comment from @Addison-Stavlo :

> Imagine a user likes a template part supplied by a theme so much that they add it to a custom template(s) and never edit the content of the template part, if auto-draft template parts were wiped on theme change then that custom template would lose part of its content.

Really, what should be happening is when a template is published, any template parts inside it should be automatically published as well. If that happens, it actually removes the need for the theme meta on both templates and template parts. That would represent option 3 as listed above.

The strongest alternative is option 2, which does not de-couple edited templates and parts from their original theme and therefore requires the meta. I personally would prefer option 3, although I am now a little bit more open to option 2 than before.

Option 4 is a terrible idea that surely no one would support. Option 1 is what I was previously supporting but you can disregard that now.